### PR TITLE
One final round of mem errors

### PR DESCRIPTION
**Pull request by ricrogz**
The final round of leak fixes (and some more). These took longer than the rest because I made the mistake of looking into the Mol Descriptor issues first, which, unfortunately, were not just leaks, but also a couple other issues on top.

In any case, here we have:
- Fixes to leaking objects in  `Wrap/Enumerate.cpp`, `Wrap/FilterCatalog.cpp` and `Wrap/rdMolDescriptors.cpp`: we need to declare those wrappers as `shared_ptr`s, and not plain pointers, or else Python won't take care of cleaning up the new objects it creates.

- in `testMolDescriptors.py`, registering the Python declared `NumAtoms` `PropertyFunctor` was overwriting the C++ functor with the same name that is initialized [here](https://github.com/rdkit/rdkit/blob/cf29f0895fbe7397c38a2e6c0b52f70ab9dd5108/Code/GraphMol/Descriptors/Property.cpp#L58). I'm not quite sure why this doesn't break the test as it is now (the C++ `NumAtoms` is called with `onlyExplicit = false`, while the Python one defaults to `true`), but the test fails if none of the other tests in the file is executed first. I don't think that overwriting the C++ functor is intentional, so I just changed the internal name of the Python one to avoid it. 
 
- in `registerProperty()`, when overriding an existing functor, we return the index of the functor in the registry, but when adding the new one we return the number of registered functors, which is off-by-one from the index of one we just added. I fixed this to always return the index. I don't think this is relevant, but it's probably good to be consistent :)

- in `RegisterDescriptor.h` I noticed that we created a bunch of subclasses of `PropertyFunctor` and create a static instance of each of them, which isn't used for anything except adding the functor to the registry. I took the liberty of refactoring this code, dropping the subclasses and the static instances. The change in `PropertyFunctor::operator()` in `Property.h` is also related to this.

- finally, there was a life time issue with  `PythonPropertyFunctor`: the `decref()` call in the destructor would only be hit when the object (coming from Python) would be in the process of being cleaned up, meaning that either its ref count is already at 0, or that Python is shutting down, and no longer cares about ref counts. Even worse: since `registerProperty()` takes ownership of the pointers it receives (stated in `Property.h`, and due to `registry` being declared as `vector<shared_ptr<PropertyFunctor>>`), the destructor will be called when `registry` itself is being destroyed, which, again, will happen when Python is shutting down. This gives us nice errors like this one:
    ```
    133: Fatal Python error: PyThreadState_Get: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
    133: Python runtime state: finalizing (tstate=0x0000612000000050)
    133: 
     ```
    To avoid such issues, I added `registerPropertyHelper()`. As the comment says, it ~creates a copy of the functor, so that the `registry` can take ownership and manage it, and it also~ increases the ref count of the original functor so that Python doesn't clean it up when it goes out of scope (we still need the Python-declared `__call__` function) . ~This way, Python can clean up the original functor when it is shutting down, and the `registry`'s destruction will take care of the copy, and none will interfere with the other one~. Then, it registers the shared pointer to the functor so that Python and the static vector share ownership.
    
    Note that removing the `incref()` from the `PythonPropertyFunctor` constructor doesn't have any adverse effects: Python will manage the lifetime of such objects. We don't really care about them being cleaned up when going out of scope unless they get registered, and then `registerPropertyHelper()` takes care of incrementing the ref count so they don't go away.

And with this, and once https://github.com/rdkit/rdkit/pull/7872 is finished and merged, the code covered by tests will practically be leak free.

(... to be continued ... but hopefully not anytime soon).

### CODE REVIEW

**Comment by ricrogz** [comment]
Almost missed a big one: `pySimDivPickers` was leaking about 4 mb of memory, lost in allocated arrays that weren't being deallocated if the functions being called did throw.

**Comment by ricrogz** [comment]
... and another one. Strangely, asan 6 did flag some leaks here, but asan 8 didn't see them.

**Comment by bp-kelley** [comment]
Good catch on the shared-ptrs wrap, I think we missed this idiom years ago.

File: Code/Numerics/Optimizer/BFGSOpt.h [hunk 1]
```diff
@@ -228,15 +212,15 @@ int minimize(unsigned int dim, double *pos, double gradTol,
...
   }
   // pick a max step size:
-  maxStep = MAXSTEP * std::max(sqrt(sum), static_cast<double>(dim));
+  double maxStep = MAXSTEP * std::max(sqrt(sum), static_cast<double>(dim));
 
-  for (unsigned int iter = 1; iter <= maxIts; iter++) {
+  for (unsigned int iter = 1; iter <= maxIts; ++iter) {
     numIters = iter;
-    int status;
+    int status = -1;
```
Comment (greglandrum) [inline Code/Numerics/Optimizer/BFGSOpt.h:219]: 👍

**Comment by greglandrum** [comment]
> Good catch on the shared-ptrs wrap, I think we missed this idiom years ago.

we certainly aren't consistent about how we use it

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
calling this the "final" round is certainly optimistic. ;-)

**Comment by ricrogz** [comment]
> calling this the "final" round is certainly optimistic. ;-)

Hence the "to be continued" at the end of the description :D`

**Pull request merged**